### ISSUE TITLE
Is it possible to run specific javascript after webview launched ?

### ISSUE THREAD

**Issue by ybdesire** [r0x0r/pywebview#123]
Is it possible to run specific javascript after webview launched ?
Such as Android evaluateJavascript(), I want to do some auto click by running javascript after webview lunched.

**Comment by r0x0r** [comment]
Wouldn't embedding javascript in the opened web page do the trick?

**Comment by ybdesire** [comment]
Thanks @r0x0r for the response. But How can we wouldn't embedding js in pywebview ? Would you share more details ?

**Comment by r0x0r** [comment]
e.g.

```
webview.create_window("Test window", "")
webview.load_html("YOUR HTML CODE <script>alert('this is executed as soon as html is loaded')</script>")
```

or embed javascript inside the SCRIPT tag in a similar manner into the html file opened with create_window. Hope that helps.

**Comment by ybdesire** [comment]
Cool.

Do we have API like get_html_from_current_webview(webview) ?
So the 'YOUR HTML CODE' can be got directly from created/loaded webview.

**Comment by shivaprsdv** [comment]
I think the OP is asking whether we can make pywebview inject
custom JS into any random page the user might open. So he has
no control over what is being loaded. This calls for an API method,
something like `webview.run_javascript(str)`. WebKit has an API
method for this on all platforms; if we could find a way on Trident,
this feature can be implemented cross-platform.

**Comment by shivaprsdv** [comment]
@ybdesire Can you please edit the title? It should read "...webview
_launched_", not "_lunched_".

**Comment by ybdesire** [comment]
@shivaprsdv Fixed. Thanks.

**Comment by r0x0r** [comment]
@ybdesire what is your use case exactly? Ie what would you need the get_html_code function for?

**Comment by shivaprsdv** [comment]
@ybdesire If you are on Mac, an `evaluate_js()` method is now
available in the new `eval-js` branch. You can run your JS code
by calling `webview.evaluate_js("<your code>")`.

**Comment by r0x0r** [comment]
I have pushed a WinForms implementation. One unsolved problem is that that in order to return a value from the JS code, the code must be wrapped inside a function and the return value must be preceded by the return keyword.  We cannot require a return keyword in the invoked code, as having it outside a function is not valid Javascript.

**Comment by r0x0r** [comment]
QT / GTK implementations are in as well. Returning a value from Javascript is still a problem. QT and Cocoa support it out of the box, but GTK and WinForms have no straightforward to do it. As I mentioned earlier WinForms could be done, by using the return keyword. QT is fine with it, but Cocoa returns undefined if the last evaluated value is prepended with return. As for GTK I could not find any solution from my limited research.

**Comment by shivaprsdv** [comment]
> As for GTK I could not find any solution from my limited research.

Same here. The problem with GTK is inherent in the native C code,
the `execute_script` function is [defined](https://webkitgtk.org/reference/webkitgtk/stable/webkitgtk-webkitwebview.html#webkit-web-view-execute-script) to return `void`, so there
is no direct way the value can be returned from the method.

A hack I've found on a blog is to modify the script to store its result
in the document's title, and then connect to the `notify::title` signal
of the webview to handle the trigger. Here's the technique:
![screen shot 2017-09-16 at 9 38 13 pm](https://user-images.githubusercontent.com/18247990/30513834-89328e46-9b27-11e7-84e3-45eb6e873bdf.png)

**Comment by r0x0r** [comment]
I came up with a solution for WinForms. By wrapping the code into
```
function() {
 return eval("code")
}
``` 
we can return a value from the last expression without resorting to the return keyword. The double quotes must be escaped in the code. 

I am not entirely sure of the ramifications of using eval here. Could there cases where it will fail? Or security concerns?

**Comment by shivaprsdv** [comment]
Questions aside, good idea. Incidentally, this is just the thing we need to make it
work on GTK as well. The technique is to store the result of the expression in the
title of the document; then we can call the `get_title()` method of the webview
to get it later. I ditched my attempt because I didn't know about `eval()`. 😇 

I've done a bit of research, seems like `eval()` would _work_ pretty well. Regarding
security, what many sources like MDN warns is the possibility of injection, since
the code is passed as a raw string; but I don't think it matters here, because we
are not sending it over a network, but within the local runtime environment.

Since there is no alternative to it, I suggest to go on with it; it's better than having
nothing. Meanwhile let me try if I can make it work on GTK.

**Comment by shivaprsdv** [comment]
I have pushed the GTK part. Worked well in my tests. Check it out.

Btw @r0x0r, with the latest rebase you seem to have messed up the
commit history with duplicate commits. 🤔

**Comment by r0x0r** [comment]
Good work. I've refactored it a bit, mainly got rid of the magic strings in the code and replaced the title change event with a custom handler.

**Comment by shivaprsdv** [comment]
Call it a day, work done. Anything left before merge?

@ybdesire Can you please test whether the new `evaluate_js()` API
does what you want?

**Comment by shivaprsdv** [comment]
There is a bug in the current implementation of `evaluate_js()` in GTK.
Since the JS code is folded into a single line string passed to `eval()`,
we cannot have // comments in the code. Doesn't WinForms have the
same problem?

IMO parsing the code to remove comments is overkill, I have checked.
If we are sticking with this, better state it in the documentation. Users
can of course use multi-line comments.

**Comment by r0x0r** [comment]
Why fold the code into a single line? eval() works just fine with line breaks.

**Comment by shivaprsdv** [comment]
No it shouldn't. What I mean is 👇 this, from my browser console:

![screen shot 2017-09-18 at 9 01 08 pm](https://user-images.githubusercontent.com/18247990/30550264-88e8e46a-9cb4-11e7-826f-337242155030.png)
See, only first line is taken as the string, without the trailing `\`.

Now I have realised, we actually need to _escape_ the line-breaks
instead of _removing_ them. Please check-out the commit I pushed,
it will become clear to you. I have replaced the /* comments */ with
// comments in the example script. Without the changes I made to
_gtk.py_, it wouldn't run properly. Let me know if does on Windows.

**Comment by r0x0r** [comment]
Yes, you are correct. I had thought about escaping line breaks, but the problem did not manifest itself, unless single line comments were used. WinForms had the same problem, but now it is fixed there as well. I refactored it a bit and wrote a test case. I think we are done here

**Comment by shivaprsdv** [comment]
@r0x0r I just found that the QT implementation of `evaluate_js()` worked only on
PyQt4. I have pushed a fix to master (672d1b592110fd192e4694b37c269ec03764a79a) to make it work on PyQt5 as well.
